Add `CowStr` enum

KushalMeghani1644 opened the pull request:
### Overview
This PR adds a `Cow<'a, N, LenT>` type to the heapless crate, enabling efficient clone-on-write strings compatible with heapless data structures.

### Features
- `Cow::Borrowed(&StringView<LenT>)` for zero-copy references.
- `Cow::Owned(String<N, LenT>)` for owned heapless strings.
- Conversion helpers: `to_owned()`, `as_str()`, `is_borrowed()`, `is_owned()`.
- `From<&StringView<LenT>>` and `From<String<N, LenT>>` for easy construction.
- Implements `Borrow<str>` for compatibility with APIs expecting `&str`.

### Motivation
Currently, heapless strings require manual cloning or reference handling. `Cow` simplifies this by providing a single type that can efficiently represent either a borrowed or owned string.

### Testing
Verified compilation and basic API usage. Unit tests for `Cow` will follow in a separate PR.

- This PR addresses issue #582

zeenix commented on src/cow.rs:
```diff
@@ -0,0 +1,85 @@
+#![cfg(feature = "alloc")]
+
+extern crate alloc;
```
The whole point of this crate is to avoid `alloc`. :)

zeenix commented on src/cow.rs:
```diff
@@ -0,0 +1,85 @@
+
+/// A clone-on-write string type for heapless
+#[derive(Debug)]
+pub enum Cow<'a, const N: usize, LenT: LenType = usize> {
```
Unlike the `std::borrow::Cow`, this one is limited to strings only. I feel that naming it as such, would give the wrong impression. I'd rather we have a different name (e.g `CowStr`).

zeenix commented on src/cow.rs:
```diff
@@ -0,0 +1,85 @@
+#[derive(Debug)]
+pub enum Cow<'a, const N: usize, LenT: LenType = usize> {
+    /// A borrowed view of a string.
+    Borrowed(&'a StringView<LenT>),
+
+    /// An owned string with inline storage of size `N`.
+    Owned(String<N, LenT>),
```
Since this is going to be string-specific, it's probably best to take advantage of that fact and also support another variant that never deep-clones (even on `to_owned`): `Static(&'static StringView<LenT>)`. This is similar to [what we added in zvariant](https://github.com/dbus2/zbus/blob/main/zvariant/src/str.rs#L34).

zeenix reviewed (changes requested):
We'll want the unit tests in the same PR, to ensure they don't get forgotten.

sgued commented:
Can you please motivate a bit more the reasons for this PR (Through a benchmark for example) ?

Here it looks like `Cow<N>` is at least as large as `String<N>` so no memory is saved, and moving/copying it will be expensive (unlike the `std` Cow where the `Borrowed` case an allocation is saved).

KushalMeghani1644 commented:
Hi @sgued,
I ran the benchmarks after the latest changes, and currently all the cow_bench.rs benches show 0 measured runs. It seems like the benchmark harness isn’t picking up any tests. (I tried to add criterion 0.7.0 to Cargo.toml to run the benches and it shows 0 benches ran) I am still trying to work on the bench to get it working, it will probably take some time, though. Thus, I haven't pushed the cow_bench.rs to this PR yet :)

Thanks for putting time to my PR

KushalMeghani1644 commented:
Hey @zeenix I have added the missing tests, sorry for the inconvenience!

sgued commented:
You still haven't justified the benefits for this. What is a use-case where the performance or memory usage here is noticeable?

KushalMeghani1644 commented:
Thanks for the review. @sgued, I am now aware that `CowStr<N>` does not reduce the size
of the value compared to `String<N>` with inline storage the `enum` is as large as
the owned variant. The purpose here is not memory footprint, but conditional
copy avoidance and API ergonomics. In embedded flows we often parse or normalize
input strings where (for the majority of calls) the input already satisfies the
required invariant. In those cases `CowStr` lets us return a borrowed view and
avoid an O(n) byte copy. Only in the minority case do we pay the cost of cloning into an owned buffer.

If you are fine with it! then, I can add a Criterion benchmark comparing:

     1. A baseline function that always copies into a new String<N>.
     2. A function returning `CowStr` that only clones on mutation.

When the ‘no-change’ path dominates, this eliminates most copying. That can reduce cycle count and energy on MCUs. If the caller ultimately always needs an owned value, then I agree CowStr doesn’t help in that situation.

If this conditional-clone use-case is not compelling for heapless itself, I’m OK
with withdrawing this PR, or making a external crate, maybe something like heapless-cow.

Thanks allot for putting time into my PR :D

sgued commented:
Please provide such a benchmark.
My concern is that even in the `Borrowed` case, moving the `CowStr` around is still done through `memcpy'ing` the entire size of the `CowStr`, thus not gaining any performance.

zeenix commented:
@KushalMeghani1644 Thanks so much for looking into extensive benchmarking. :+1: I have to admit that I'm having a bit of difficulty understanding what exactly is being compared. I think it would be much easier to measure and see the actual benefits (if any) of this PR, if you could add the benchmarks that only use the existing API and then modified to make use of `CowStr`. Could you modify the PR to be 2-3 commits, so the first commit adds the benchmarks, the second one adds the new API and the last commit modifies the benchmarks? That way it would be super easy to reproduce the results locally and easier to read and verify the benchmark code for us. :pray: 

BTW, please do not open a new PR for this. Just force push to this branch.

KushalMeghani1644 commented:
Hey @zeenix I have changed everything as you requested:

**Changes Made:**

1. Fixed the nested match statement issue by refactoring the bench_cowstr_to_owned function to use a helper function with const generics (bench_cowstr_to_owned_for_size<const N: usize>), eliminating the unreachable code and nested matching pattern.

2. Restructured the commit history into three clean commits as requested:

Commit 1: "Add baseline benchmarks for String operations" - Contains benchmarks using only the existing String API
without any CowStr code

Commit 2: "Add CowStr clone-on-write string type" - Adds the complete CowStr implementation with tests

Commit 3: "Add CowStr benchmarks comparing against String baseline" - Adds comprehensive CowStr benchmarks that can be compared against the baseline

This structure makes it easy to run benchmarks before and after the CowStr
implementation, clearly showing the performance benefits of the new API.

3. Force pushed to the existing branch as requested, replacing the messy commit
history with a clean, logical progression that's easy to review and reproduce
locally.

Both benchmark files compile successfully and all tests pass. The benchmarks can now be run to compare String cloning performance against CowStr's clone-on-write behavior.

Thanks a lot for putting time to my PR :)

zeenix reviewed (changes requested):
Thank you for updating the PR. However, you're adding new benchmarks for the CowStr in the last commit. I asked for modifying the benchmarks to make use of the CowStr, so it's easier to see/observe the difference.

Moreover, we have a suspicion that you're using AI tools for not only the coding part but also writing your comments. While I have no objections on making use of AI tools for the code (as long as it's good quality code), I think using AI to reply to us, is unacceptable. Please ensure that your following comments on this PR are written exclusively by you.

KushalMeghani1644 commented:
Hey @zeenix I get your concerns regarding the use of AI tools, but just to clearify myself, I used AI ONLY for helping me make the bench! and the comments and the commit messages, all were by my own! and as you said:

"While I have no objections on making use of AI tools for the code (as long as it's good quality code), I think using AI to reply to us, is unacceptable. Please ensure that your following comments on this PR are written exclusively by you."

I had manually reviewed the codes AI was suggesting me before implementing, and as of the comments/messages they are indeed EXCLUSIVELY written by me.

 If you want instead of having an entirely new benchmark I can indeed make changes in the old existing one aswell, I had made a new bench since my local clone of repo was getting messier :) I'll appreciate your response.

KushalMeghani1644 commented:
>I'm sorry but I find that hard to believe, given that the wording of your comment https://github.com/rust-embedded/heapless/pull/615#issuecomment-3368818154 matches the style of Claude 100%. But ok, let's move on.. 

I am sorry but this is kinda harsh :) I don't even use Claude anymore I use GPT-5 for helping me for code assistance xD and as I stated, I don't use AI for commenting on whatever you ask me to do, but as you said, let's move on :) 

So just to clearify you need me to make a bench that doesn't uses the `CowStr` enum I intend to add in this PR and then another commit after that which modifies that existing bench to use `CowStr` so you can clearly see the differences, sorry if I am taking too long but to be a bit more perfect I need a bit of clearity. 

Thanks

KushalMeghani1644 commented:
Hey @zeenix I have done everything as you asked, with a clear commit history which adds a benchmark for comparing my changes before vs after ;) thanks for helping me, I am sorry to you if I was harsh on my previous comments, and if I took too long for this.

zeenix commented:
> I have done everything as you asked, with a clear commit history which adds a benchmark for comparing my changes before vs after ;)

Not really. The last commit still mostly adds new benchmarks, instead of **only** modifying existing ones. Moreover, you still merged master instead of rebasing on it (as I asked). To make things worse, you're completely ignoring the CI being red.
